### PR TITLE
Gracefully handle task requests

### PR DESCRIPTION
**Pull request by Ladicle**

### CODE REVIEW

File: engine/internal/processor/processor.go [hunk 1]
```diff
@@ -174,32 +159,48 @@ func (p *P) NeedLeaderElection() bool {
...
-	}()
+		defer func() { _ = stream.CloseSend() }()
 
-	go func() {
-		if err := p.processTasks(ctx, stream); err != nil {
-			errCh <- fmt.Errorf("process tasks: %s", err)
+		eg, ctx := errgroup.WithContext(ctx)
+		eg.Go(func() error { return p.sendEngineStatusPeriodically(ctx, stream) })
+		eg.Go(func() error { return p.processTasks(ctx, stream) })
+		return eg.Wait()
```
Comment (kkaneda) [inline engine/internal/processor/processor.go:186]: Supposet that sendEngineStatusPeriodically() returns an error. `eg.Wait` does not return the error until `processTasks` completes IIUC. Does this work when `inference-manager-server` becomes unavailable for a short period of time?

File: engine/internal/processor/processor.go [hunk 2]
```diff
@@ -285,9 +313,26 @@ func (p *P) processTask(
 	if err := p.modelSyncer.PullModel(ctx, taskModel(t)); err != nil {
 		return fmt.Errorf("pull model: %s", err)
 	}
-	// TODO(aya): Consider how to handle the case where the runtime is scaled down before the request is sent.
 
-	req, err := p.buildRequest(ctx, t)
+	done := make(chan string)
```
Comment (kkaneda) [inline engine/internal/processor/processor.go:317]: I guess `atomic.Bool` is another option? No opinion on either way, though.

File: engine/internal/processor/processor.go [hunk 3]
```diff
@@ -174,32 +159,48 @@ func (p *P) NeedLeaderElection() bool {
...
-
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
 
-	errCh := make(chan error, 2)
-	go func() {
-		if err := p.sendEngineStatusPeriodically(ctx, stream); err != nil {
-			errCh <- fmt.Errorf("send engine status: %s", err)
+	run := func() error {
+		streamCtx, streamCancel := context.WithCancel(context.Background())
```
Comment (kkaneda) [inline engine/internal/processor/processor.go:172]: Out of of scope of this PR, but can we make `inference-manager-engine` stop receiving new tasks when grace period starts?

**Review by kkaneda — APPROVED** [review]
LG!

File: engine/internal/processor/processor.go (hunk 1 above)
Comment (Ladicle) [reply engine/internal/processor/processor.go:186]: Yes, in that case, `stream.Recv()` fails, and `processTasks` returns an error.

File: engine/internal/processor/processor.go (hunk 3 above)
Comment (kenji-cloudnatix) [reply engine/internal/processor/processor.go:172]: got it!

**Pull request merged**